### PR TITLE
pfe-select Component

### PR DESCRIPTION
**Pull request by diwanshi**
## Title of this pull request

* Link to discovery documentation (should contain all the planning and requirements for this work):
   - See the `DISCOVERY.md` template at the root of the project
* Link(s) to demo pages where this element can be viewed:

---

### What has changed and why
Summarize files edited as part of this MR along with a brief description of what was changed/why.

*

### Testing instructions
Be sure to include detailed instructions on how your update can be tested by another developer.

1. 

#### Browser requirements
Your component should work in all of the following environments:

- [ ] Latest 2 versions of Edge
- [ ] Internet Explorer 11 (should be useable, not pixel perfect)
- [x] Latest 2 versions of Firefox (one on Mac OS, one of Windows OS)
- [x] Firefox 60.7.2 (or latest version for Red Hat Enterprise Linux distribution)
- [x] Latest 2 versions of Chrome (one on Mac OS, one of Windows OS)
- [x] Latest 2 versions of Safari
- [ ] Galaxy S9 Firefox
- [ ] iPhone X Safari
- [ ] iPad Pro Safari
- [ ] Pixel 3 Chrome

Your repository infrastructure updates should work for at least:
- [ ] Node v8.x
- [ ] NPM v7.x

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one feature request or issue (no stragglers)?
- [x] Did browser testing pass?
- [x] Did you update or add any necessary documentation (README.md, WHY.md, etc.)?
- [x] Was this feature demo'd and the design review approved?
- [ ] Did you update the CHANGELOG.md file with a summary of this update?

**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**

### CODE REVIEW

File: elements/pfe-select/demo/index.html [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    <script src="/@webcomponents/webcomponentsjs/webcomponents-loader.js"></script>
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.5/require.min.js"></script>
+    <script>require(['../pfe-select.umd.js'])</script>
+  </head>
+  <body unresolved>
+
+    <h1>&lt;pfe-select&gt;</h1>
+    <label for="example1">Example one</label>
+    <pfe-select id="pfe-select-success">
+      <select slot="pfe-select" id="example1">
```
Comment (diwanshi) [inline elements/pfe-select/demo/index.html]: Remove the name of the slot

File: elements/pfe-select/src/pfe-select.html [hunk 2]
```diff
@@ -0,0 +1 @@
+<slot name="pfe-select"></slot>
```
Comment (diwanshi) [inline elements/pfe-select/src/pfe-select.html]: remove the name of the slot

File: elements/pfe-select/src/pfe-select.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    this.querySelector('select').setAttribute('aria-invalid', invalidAttr);
+  }
+
+  static get observedAttributes() {
+    return ["pfe-invalid"];
+  }
+
+  constructor() {
+    super(PfeSelect);
+    this._pfeOptions;
```
Comment (diwanshi) [inline elements/pfe-select/src/pfe-select.js]: initialize as null

File: elements/pfe-select/src/pfe-select.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    attr = attr.replace("pfe-", "");
+    // If the observer is defined in the attribute properties
+    if (this[attr] && this[attr].observer) {
+      // Get the observer function
+      let observer = this[this[attr].observer].bind(this);
+      // If it's a function, allow it to run
+      if (typeof observer === "function") observer(attr, oldValue, newValue);
+    }
+
+    if (this.pfeInvalid) {
```
Comment (diwanshi) [inline elements/pfe-select/src/pfe-select.js]: update the pfeInvalid value with observer's value

File: elements/pfe-select/src/pfe-select.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    });
+  }
+
+  modifyDOM() {
+    // Create select element
+    let pfeSelect = document.createElement('select');
+    pfeSelect.setAttribute('slot', 'pfe-select');
+    // Create option element for each element in _pfeOptions array
+    this._pfeOptions.map(el => {
+      const option = Object.assign(document.createElement('option') , el);
```
Comment (diwanshi) [inline elements/pfe-select/src/pfe-select.js]: This needs a polyfill

Comment (kylebuch8) [reply elements/pfe-select/src/pfe-select.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#Polyfill

File: elements/pfe-select/src/pfe-select.story.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { storiesOf } from "@storybook/polymer";
+import "../pfe-select";
+
+storiesOf("Select", module).add(
+  "pfe-select",
+  () => `
+  <pfe-select>
```
Comment (diwanshi) [inline elements/pfe-select/src/pfe-select.story.js]: update it with multiple examples

File: elements/pfe-select/src/pfe-select.scss [hunk 7]
```diff
@@ -0,0 +1,9 @@
+@import "../../pfe-sass/pfe-sass";
+
+:host {
+  display: block;
+}
+
+:host([hidden]) {
+  display: none;
+}
```
Comment (starryeyez024) [inline elements/pfe-select/src/pfe-select.scss]: @markcaron @diwanshi the lightdom CSS file is temporary; as soon as Edge releases the latest version (with Chromium under the hood) and we drop support for IE11, we can stop loading the lightdom.css files entirely. 

Can we move most of those styles back into the original web component sass file?

Comment (markcaron) [reply elements/pfe-select/src/pfe-select.scss]: Unless I'm misunderstanding...We want light DOM support still. PFE Select is intended to style all standard selects, including forms that may or may not require JavaScript.

Or is everyone else agreeing that all non-JS support should look like browser-chromed inputs? I, personally, feel that we should try to ship as much as possible without JS. I know this means there's an extra include (until we get a service to do all this for the Dev).

File: elements/pfe-select/src/pfe-select.js (hunk 5 above)
Comment (diwanshi) [reply elements/pfe-select/src/pfe-select.js]: I've added the polyfill from the mdn doc.

File: elements/pfe-select/src/pfe-select.scss (hunk 7 above)
Comment (diwanshi) [reply elements/pfe-select/src/pfe-select.scss]: Hi @markcaron 

I moved back all the styles in lightdom file to this file. Can you please review if everything is okay. I tested this and it works fine.

**Comment by diwanshi** [comment]
@markcaron ,

While I was testing this component on both firefox and chrome I noticed that there are keyboard accessibility issue in chrome. 
If I use Tab it focus the element, when I press down arrow key it doesn't open the options list and if I enter then it selects the first/default value. 
So the options list is not opening but if I keep pressing down and enter it shows me the selected value. 

Can you confirm if this is happening on your end too? If possible can you give me a pointer on how to solve this.

**Comment by markcaron** [comment]
@diwanshi, I'll look into it. Thanks!

**Comment by markcaron** [comment]
I'm not seeing any keyboard access issues in FF or Chrome. They're behaving as expected for me.

However, they're not working at all in Safari. In fact, it locks up the browser! I'll investigate.

**Comment by markcaron** [comment]
This seems to be an issue with my current version of Safari (13.0.1). BrowserStack's Safari 12.1.1 seems to work fine. Can we get another pair of eyes on this? @kylebuch8

File: elements/pfe-select/demo/pfe-select.story.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
+import { storiesOf } from "@storybook/polymer";
```
Comment (diwanshi) [inline elements/pfe-select/demo/pfe-select.story.js]: @kylebuch8 hey I've added storybook as well. Please let me know if this is okay and covers everything.
Also, I'm facing an issue with the markup on the storybook HTML. The tags are not indented and I'm not sure why that might be happening. Can you take a look and suggest something?

**Comment by kylebuch8** [comment]
@diwanshi I can't get any of the styles to compile. This is the error that I'm getting.

```
@patternfly/pfe-select: Error in plugin "sass"
@patternfly/pfe-select: Message:
@patternfly/pfe-select:     ../pfe-sass/pfe-sass.scss
@patternfly/pfe-select: Error: It's not clear which file to import for '@import "_functions"'.
@patternfly/pfe-select:        Candidates:
@patternfly/pfe-select:          __functions.scss
@patternfly/pfe-select:          _functions.scss
@patternfly/pfe-select:        Please delete or rename all but one of these files.
@patternfly/pfe-select:         on line 28 of ../pfe-sass/pfe-sass.scss
@patternfly/pfe-select:         from line 1 of src/pfe-select.scss
@patternfly/pfe-select: >> @import "_functions";
@patternfly/pfe-select:    ^

```

**Comment by kylebuch8** [comment]
> @diwanshi I can't get any of the styles to compile. This is the error that I'm getting.
> 
> ```
> @patternfly/pfe-select: Error in plugin "sass"
> @patternfly/pfe-select: Message:
> @patternfly/pfe-select:     ../pfe-sass/pfe-sass.scss
> @patternfly/pfe-select: Error: It's not clear which file to import for '@import "_functions"'.
> @patternfly/pfe-select:        Candidates:
> @patternfly/pfe-select:          __functions.scss
> @patternfly/pfe-select:          _functions.scss
> @patternfly/pfe-select:        Please delete or rename all but one of these files.
> @patternfly/pfe-select:         on line 28 of ../pfe-sass/pfe-sass.scss
> @patternfly/pfe-select:         from line 1 of src/pfe-select.scss
> @patternfly/pfe-select: >> @import "_functions";
> @patternfly/pfe-select:    ^
> ```

Never mind. I was able to get things working locally.

File: elements/pfe-select/demo/pfe-select.story.js (hunk 8 above)
Comment (diwanshi) [reply elements/pfe-select/demo/pfe-select.story.js]: @kylebuch8 i've updated the code for storybook and have added read me files as we discussed. The markup issue is still not resolved for me.

Comment (kylebuch8) [reply elements/pfe-select/demo/pfe-select.story.js]: @diwanshi - I think the markup issue is fine for now.

**Comment by kylebuch8** [comment]
> This seems to be an issue with my current version of Safari (13.0.1). BrowserStack's Safari 12.1.1 seems to work fine. Can we get another pair of eyes on this? @kylebuch8

@markcaron I found out that this the issue is a bug in Safari.
https://bugs.webkit.org/show_bug.cgi?id=202198

We get around the issue by adding `text-rendering: auto !important;` on the select element in the slot.

**Comment by diwanshi** [comment]
@kylebuch8 hey I rebased the branch with master and pushed it. Is there anything that I should add in pfe-select?

**Comment by kylebuch8** [comment]
@diwanshi I think we're good now. I'm going to go ahead and get this merged into master.

**Pull request merged**